### PR TITLE
Add configuration to setup collector logs via config file.

### PR DESCRIPTION
**Pull request by bogdandrutu**
Updates: https://github.com/open-telemetry/opentelemetry-collector/issues/3957

Next PRs will use this configuration and it will deprecate/remove the flags.

Signed-off-by: Bogdan Drutu <bogdandrutu@gmail.com>

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -153,13 +153,38 @@ func (cfg *Config) validateServicePipelines() error {
...
 
+// ServiceTelemetry defines the configurable settings for service telemetry.
+type ServiceTelemetry struct {
+	Logs ServiceTelemetryLogs `mapstructure:"logs"`
+}
+
+// ServiceTelemetryLogs defines the configurable settings for service telemetry logs.
+// This MUST be compatible with zap.Config. Cannot use directly zap.Config because
+// the collector uses mapstructure and not yaml tags.
+type ServiceTelemetryLogs struct {
```
Comment (mx-psi) [inline config/config.go:173]: Maybe check in `Validate` that the `Level` and `Encoding` fields have a valid value?

Comment (bogdandrutu) [reply config/config.go:174]: Nice suggestion, done.

**Review by mx-psi — APPROVED** [review]

**Comment by bogdandrutu** [comment]
Please review @Aneurysm9

File: config/configunmarshaler/testdata/valid-config.yaml [hunk 2]
```diff
@@ -20,6 +20,11 @@ extensions:
     extra: "some string"
 
 service:
+  telemetry:
```
Comment (jpkrohling) [inline config/configunmarshaler/testdata/valid-config.yaml:23]: Would you be able to add an example of what's expected to happen when this is used, and what's expected when "json" encoding is used? I tried this PR here locally, they look the same:

```
extensions:
  zpages:

receivers:
  otlp:
    protocols:
      grpc:

processors:

exporters:
  logging:

service:
  telemetry:
    logs:
      level: "DEBUG"
      development: true
      encoding: "json"
  extensions: [zpages]
  pipelines:
    traces:
      receivers:
        - otlp
      processors: []
      exporters:
        - logging
```

With this output:

```
$ go run ./cmd/otelcol/ --config local/simple.yaml 
2021-09-15T10:19:06.664+0200	info	service/collector.go:255	Starting otelcol...	{"Version": "latest", "NumCPU": 8}
2021-09-15T10:19:06.665+0200	info	builder/exporters_builder.go:266	Exporter was built.	{"kind": "exporter", "name": "logging"}
2021-09-15T10:19:06.665+0200	info	builder/pipelines_builder.go:216	Pipeline was built.	{"pipeline_name": "traces", "pipeline_datatype": "traces"}
2021-09-15T10:19:06.665+0200	info	builder/receivers_builder.go:229	Receiver was built.	{"kind": "receiver", "name": "otlp", "datatype": "traces"}
2021-09-15T10:19:06.665+0200	info	service/service.go:93	Starting extensions...
2021-09-15T10:19:06.665+0200	info	builder/extensions_builder.go:54	Extension is starting...	{"kind": "extension", "name": "zpages"}
2021-09-15T10:19:06.665+0200	info	zpagesextension/zpagesextension.go:40	Register Host's zPages	{"kind": "extension", "name": "zpages"}
2021-09-15T10:19:06.666+0200	info	zpagesextension/zpagesextension.go:53	Starting zPages extension	{"kind": "extension", "name": "zpages", "config": {"TCPAddr":{"Endpoint":"localhost:55679"}}}
2021-09-15T10:19:06.666+0200	info	builder/extensions_builder.go:60	Extension started.	{"kind": "extension", "name": "zpages"}
2021-09-15T10:19:06.666+0200	info	service/service.go:98	Starting exporters...
2021-09-15T10:19:06.666+0200	info	builder/exporters_builder.go:93	Exporter is starting...	{"kind": "exporter", "name": "logging"}
2021-09-15T10:19:06.666+0200	info	builder/exporters_builder.go:98	Exporter started.	{"kind": "exporter", "name": "logging"}
2021-09-15T10:19:06.666+0200	info	service/service.go:103	Starting processors...
2021-09-15T10:19:06.666+0200	info	builder/pipelines_builder.go:52	Pipeline is starting...	{"pipeline_name": "traces", "pipeline_datatype": "traces"}
2021-09-15T10:19:06.666+0200	info	builder/pipelines_builder.go:63	Pipeline is started.	{"pipeline_name": "traces", "pipeline_datatype": "traces"}
2021-09-15T10:19:06.666+0200	info	service/service.go:108	Starting receivers...
2021-09-15T10:19:06.666+0200	info	builder/receivers_builder.go:71	Receiver is starting...	{"kind": "receiver", "name": "otlp"}
2021-09-15T10:19:06.666+0200	info	otlpreceiver/otlp.go:74	Starting GRPC server on endpoint 0.0.0.0:4317	{"kind": "receiver", "name": "otlp"}
2021-09-15T10:19:06.666+0200	info	builder/receivers_builder.go:76	Receiver started.	{"kind": "receiver", "name": "otlp"}
2021-09-15T10:19:06.666+0200	info	service/telemetry.go:65	Setting up own telemetry...
2021-09-15T10:19:06.679+0200	info	service/telemetry.go:113	Serving Prometheus metrics	{"address": ":8888", "level": 0, "service.instance.id": "650fdd8a-9b38-4e1f-a982-2646b651c7c0"}
2021-09-15T10:19:06.679+0200	info	service/collector.go:177	Everything is ready. Begin running and processing data.
^X^C2021-09-15T10:19:09.616+0200	info	service/collector.go:191	Received signal from OS	{"signal": "interrupt"}
2021-09-15T10:19:09.616+0200	info	service/collector.go:278	Starting shutdown...
2021-09-15T10:19:09.616+0200	info	service/service.go:128	Stopping receivers...
2021-09-15T10:19:09.616+0200	info	service/service.go:133	Stopping processors...
2021-09-15T10:19:09.616+0200	info	builder/pipelines_builder.go:71	Pipeline is shutting down...	{"pipeline_name": "traces", "pipeline_datatype": "traces"}
2021-09-15T10:19:09.616+0200	info	builder/pipelines_builder.go:77	Pipeline is shutdown.	{"pipeline_name": "traces", "pipeline_datatype": "traces"}
2021-09-15T10:19:09.616+0200	info	service/service.go:138	Stopping exporters...
2021-09-15T10:19:09.616+0200	info	service/service.go:143	Stopping extensions...
2021-09-15T10:19:09.616+0200	info	service/collector.go:294	Shutdown complete.
```

Comment (mx-psi) [reply config/configunmarshaler/testdata/valid-config.yaml:23]: IIUC that will happen on #4011

**Review by jpkrohling — APPROVED** [review]

File: config/config.go [hunk 3]
```diff
@@ -84,35 +86,30 @@ func (cfg *Config) Validate() error {
...
-func (cfg *Config) validateServicePipelines() error {
 	// Must have at least one pipeline.
 	if len(cfg.Service.Pipelines) == 0 {
 		return errMissingServicePipelines
 	}
 
-	// Validate pipelines.
+	// Check that all pipelines have at least one receiver and one exporter, and they reference
+	// only configured components.
 	for _, pipeline := range cfg.Service.Pipelines {
```
Comment (tigrannajaryan) [inline config/config.go:113]: These changes seem unrelated to the PR topic. Maybe do in a separate PR.

File: config/config.go [hunk 4]
```diff
@@ -153,13 +150,47 @@ func (cfg *Config) validateServicePipelines() error {
...
+// ServiceTelemetryLogs defines the configurable settings for service telemetry logs.
+// This MUST be compatible with zap.Config. Cannot use directly zap.Config because
+// the collector uses mapstructure and not yaml tags.
+type ServiceTelemetryLogs struct {
+	// Level is the minimum enabled logging level.
+	Level zapcore.Level
+
+	// Development puts the logger in development mode, which changes the
+	// behavior of DPanicLevel and takes stacktraces more liberally.
+	Development bool
```
Comment (tigrannajaryan) [inline config/config.go:180]: Does this also enable logging at "Debug" level? If not then this is not equivalent to existing `--log-profile` command line option which may make the transition more difficult.

**Review by tigrannajaryan — COMMENTED** [review]
What's the depreciation plan for exiting command line flags log-level, log-profile, log-format?

Will they remain for now and if yes if both command line flags and config settings are specified which ones take the precedent?

**Comment by alolita** [comment]
@tigrannajaryan @Aneurysm9 please review asap. This is a P0 issue and we need to get this merged. Thx!!!

File: config/config.go (hunk 4 above)
Comment (Aneurysm9) [reply config/config.go:180]: Does it need to be handled with a single config option?  If a combination of two options that can be set together replace a flag that seems fine to me.

**Review by Aneurysm9 — APPROVED** [review]

**Comment by tigrannajaryan** [comment]
> @tigrannajaryan @Aneurysm9 please review asap. This is a P0 issue and we need to get this merged. Thx!!!

I already reviewed. Not sure what's the urgency. This appears to be a backwards compatible change.

**Review by anuraaga — APPROVED** [review]

File: config/config.go (hunk 3 above)
Comment (bogdandrutu) [reply config/config.go:113]: I restructure the validate to be able to add the validate for the encoding around the same place with the other validation for the service.

File: config/config.go (hunk 4 above)
Comment (bogdandrutu) [reply config/config.go:180]: indeed is a combination of two to replace the flag, but when handle the deprecated flag we will do the right thing.
